src/axom/slam/BivariateMap.hpp

```diff
@@ -750,6 +620,202 @@ template <typename T, typename BSet, typename IndPol, typename StrPol, typename
+  {
+    return m_map->flatValue(m_bsetIterator.flatIndex(), compIndex());
+  }
+
+  AXOM_HOST_DEVICE pointer operator->() const { return &(*this); }
+
+  /**
+     * \brief return the current iterator's first index into the BivariateSet
+     */
+  PositionType firstIndex() const { return m_bsetIterator.firstIndex(); }
+
+  /**
+     * \brief return the current iterator's second index (DenseIndex)
+     *        into the BivariateSet
+     */
+  PositionType secondIndex() const { return m_bsetIterator.secondIndex(); }
+
+  /// \brief return the current iterator's component index
+  PositionType compIndex() const { return this->m_pos % numComp(); }
+
+  /** \brief Returns the number of components per element in the map. */
+  AXOM_HOST_DEVICE PositionType numComp() const { return m_map->numComp(); }
+
+protected:
+  AXOM_HOST_DEVICE void advance(IndexType n)
+  {
+    this->m_pos += n;
+    // Advance associated bset iterator.
+    auto oldBsetIndex = m_bsetIterator.flatIndex();
+    auto newBsetIndex = this->m_pos / numComp();
+    m_bsetIterator += (newBsetIndex - oldBsetIndex);
+  }
+
+private:
+  BivariateMapPtr m_map;
+  typename BivariateSetType::IteratorType m_bsetIterator;
+};
+
+/**
+ * \class BivariateMap::RangeIterator
+ *
+ * \brief An iterator type for a BivariateMap, iterating over elements in an
+ *  associated BivariateSet.
+ *
+ *  Unlike the FlatIterator, which iterates over all map elements, the
+ *  RangeIterator may point to a range of elements in the case of non-unit
+ *  stride.
+ */
+template <typename T, typename BSet, typename IndPol, typename StrPol, typename IfacePol>
+template <bool Const>
+class BivariateMap<T, BSet, IndPol, StrPol, IfacePol>::RangeIterator
+  : public IteratorBase<RangeIterator<Const>, SetPosition>
+{
+public:
+  using IterBase = IteratorBase<RangeIterator<Const>, SetPosition>;
+
+  using MapIterator = typename MapType::template MapRangeIterator<Const>;
+
+  using iterator_category = std::random_access_iterator_tag;
+  using value_type = typename MapIterator::value_type;
+  using reference = typename MapIterator::reference;
+  using pointer = typename MapIterator::pointer;
+  using difference_type = SetPosition;
+
+public:
+  using DataRefType = typename MapIterator::DataRefType;
+  using BivariateMapPtr =
+    std::conditional_t<Const, const BivariateMap*, BivariateMap*>;
+
+  using PositionType = SetPosition;
+  static constexpr PositionType INVALID_POS = -2;
+
+public:
+  /**
+     * \brief Construct a new BivariateMap Iterator given an ElementFlatIndex
+     */
+  RangeIterator(BivariateMapPtr sMap, PositionType pos)
+    : IterBase(pos)
+    , m_map(sMap)
+    , m_mapIterator(m_map->getMap()->set_begin() + pos)
+    , m_bsetIterator(m_map->set(), pos)
+  { }
+
+  /**
+     * \brief Returns the current iterator value. If the BivariateMap has
+     *        multiple components, this will return the first component.
+     *        To access the other components, use iter(comp)
```

Comment:
I don't understand the last sentence.  I can't connect `comp` to this context.

Reply:
Updated documentation to clarify.